agent/src/jmh/java/com/code_intelligence/jazzer/runtime/FuzzerCallbacksBenchmark.java

```diff
@@ -0,0 +1,216 @@
+    FuzzerCallbacks.traceCmpInt(state.arg1, state.arg2, state.pc);
+  }
+
+  @Benchmark
+  public void traceCmpIntWithPc(TraceCmpIntState state) {
+    FuzzerCallbacksWithPc.traceCmpInt(state.arg1, state.arg2, state.pc);
+  }
+
+  @Benchmark
+  @Fork(jvmArgsAppend = {"-XX:+CriticalJNINatives"})
+  public void traceCmpIntOptimizedCritical(TraceCmpIntState state) {
+    FuzzerCallbacksOptimizedCritical.traceCmpInt(state.arg1, state.arg2, state.pc);
+  }
+
+  //  @Benchmark
+  //  @Fork(jvmArgsAppend = {"--enable-native-access=ALL-UNNAMED", "--add-modules",
+  //            "jdk.incubator.foreign"})
+  //  public void
+  //  traceCmpIntPanama(TraceCmpIntState state) throws Throwable {
+  //    FuzzerCallbacksPanama.traceCmpInt(state.arg1, state.arg2, state.pc);
+  //  }
+
+  @State(Scope.Benchmark)
+  public static class TraceSwitchState {
+    @Param({"5", "10"}) int numCases;
+
+    long val;
+    long[] cases;
+    int pc = 0x12345678;
+
+    @Setup
+    public void setup() {
+      cases = new long[2 + numCases];
+      Random random = ThreadLocalRandom.current();
+      Arrays.setAll(cases, i -> {
+        switch (i) {
+          case 0:
+            return numCases;
+          case 1:
+            return 32;
+          default:
+            return random.nextInt();
+        }
+      });
+      Arrays.sort(cases, 2, cases.length);
+      val = random.nextInt();
+    }
+  }
+
+  @Benchmark
+  public void traceSwitch(TraceSwitchState state) {
+    FuzzerCallbacks.traceSwitch(state.val, state.cases, state.pc);
+  }
+
+  @Benchmark
+  public void traceSwitchWithPc(TraceSwitchState state) {
+    FuzzerCallbacksWithPc.traceSwitch(state.val, state.cases, state.pc);
+  }
+
+  @Benchmark
+  @Fork(jvmArgsAppend = {"-XX:+CriticalJNINatives"})
+  public void traceSwitchOptimizedCritical(TraceSwitchState state) {
+    FuzzerCallbacksOptimizedCritical.traceSwitch(state.val, state.cases, state.pc);
+  }
+
+  @Benchmark
+  public void traceSwitchOptimizedNonCritical(TraceSwitchState state) {
+    FuzzerCallbacksOptimizedNonCritical.traceSwitch(state.val, state.cases, state.pc);
+  }
+
+  //  @Benchmark
+  //  @Fork(jvmArgsAppend = {"--enable-native-access=ALL-UNNAMED", "--add-modules",
+  //            "jdk.incubator.foreign"})
+  //  public void
+  //  traceCmpSwitchPanama(TraceSwitchState state) throws Throwable {
+  //    FuzzerCallbacksPanama.traceCmpSwitch(state.val, state.cases, state.pc);
+  //  }
+
+  @State(Scope.Benchmark)
+  public static class TraceMemcmpState {
+    @Param({"10", "100", "1000"}) int length;
+
+    byte[] array1;
+    byte[] array2;
+    int pc = 0x12345678;
+
+    @Setup
+    public void setup() {
+      array1 = new byte[length];
+      array2 = new byte[length];
+
+      Random random = ThreadLocalRandom.current();
+      random.nextBytes(array1);
+      random.nextBytes(array2);
+      // Make the arrays agree unil the midpoint.
+      System.arraycopy(array1, 0, array2, 0, length / 2);
```

Comment:
It's just meant to represent the "average" case of an interesting memcmp. Since this is not strcmp, I don't really expect the runtime of the hook to depend on the values, but the benchmark also shouldn't assume too much about the particular implementation of the hook.

Reply:
Equal starts of bytes determine how many iterations are necessary inside the hook to compute the Hamming distance which is used to compute the index inside the value profile map.

I agree that the benchmark should not assume too much about the implementation.